lighthouse-cli/bin.ts

```diff
@@ -0,0 +1,251 @@
+if (cli.listTraceCategories) {
+  Commands.ListTraceCategories();
+}
+
+const urls = cli._;
+const outputMode = cli.output;
+const outputPath = cli['output-path'];
+const flags = cli;
+
+let config = null;
+if (cli.configPath) {
+  // Resolve the config file path relative to where cli was called.
+  cli.configPath = path.resolve(process.cwd(), cli.configPath);
+  config = require(cli.configPath);
+} else if (cli.perf) {
+  config = perfOnlyConfig;
+}
+
+// set logging preferences
+flags.logLevel = 'info';
+if (cli.verbose) {
+  flags.logLevel = 'verbose';
+} else if (cli.quiet) {
+  flags.logLevel = 'error';
+}
+
+log.setLevel(flags.logLevel);
+
+const cleanup = {
+  fns: [],
+  register(fn) {
+    this.fns.push(fn);
+  },
+  doCleanup() {
+    return Promise.all(this.fns.map(c => c()));
+  }
+};
+
+function launchChromeAndRun(addresses: Array<string>) {
+  const launcher = new ChromeLauncher({
+    autoSelectChrome: !cli.selectChrome,
+  });
+
+  cleanup.register(() => launcher.kill());
+
+  return launcher
+    .isDebuggerReady()
+    .catch(() => {
+      log.log('Lighthouse CLI', 'Launching Chrome...');
+      return launcher.run();
+    })
+    .then(() => lighthouseRun(addresses))
+    .then(() => launcher.kill());
+}
+
+function lighthouseRun(addresses) {
+  // Process URLs once at a time
+  const address = addresses.shift();
+  if (!address) {
+    return;
+  }
+
+  return lighthouse(address, flags, config)
+    .then(results => Printer.write(results, outputMode, outputPath))
+    .then(results => {
+      if (outputMode === Printer.OutputMode.pretty) {
+        const filename = `./${assetSaver.getFilenamePrefix({url: address})}.report.html`
+        Printer.write(results, 'html', filename);
+      }
+
+      return lighthouseRun(addresses);
+    });
+}
+
+function showConnectionError() {
+  console.error('Unable to connect to Chrome');
+  console.error(
+    'If you\'re using lighthouse with --skip-autolaunch, ' +
+    'make sure you\'re running some other Chrome with a debugger.'
+  );
+  process.exit(_RUNTIME_ERROR_CODE);
+}
+
+function showRuntimeError(err) {
+  console.error('Runtime error encountered:', err);
+  console.error(err.stack);
+  process.exit(_RUNTIME_ERROR_CODE);
+}
+
+function handleError(err) {
+  if (err.code === 'ECONNREFUSED') {
+    showConnectionError();
+  } else {
+    showRuntimeError(err);
+  }
+}
+
+function run() {
+  if (cli.skipAutolaunch) {
+    lighthouseRun(urls).catch(handleError);
+  } else {
+    // because you can't cancel a promise yet
+    const isSigint = new Promise((resolve, reject) => {
+      process.on(_SIGINT, () => reject(_SIGINT));
+    });
+
+    Promise
+      .race([launchChromeAndRun(urls), isSigint])
+      .catch(maybeSigint => {
+        if (maybeSigint === _SIGINT) {
+          return cleanup
+            .doCleanup()
+            .catch(err => {
+              console.error(err);
+              console.error(err.stack);
+            }).then(() => process.exit(_ERROR_EXIT_CODE));
```

Comment:
since we were always exiting with 130, and only in the case of errors logging, i refactored the code to DRY it up